master/internal/api_generic_tasks.go

```diff
@@ -0,0 +1,253 @@
+		projectID = int(*req.ProjectId)
+	} else {
+		projectID = model.DefaultProjectID
+	}
+
+	if err := a.canCreateGenericTask(ctx, projectID); err != nil {
+		return nil, err
+	}
+
+	genericTaskSpec, warnings, contextDirectoryBytes, err := a.getGenericTaskLaunchParameters(
+		ctx, req.ContextDirectory, req.Config, projectID,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := check.Validate(genericTaskSpec.GenericTaskConfig); err != nil {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"invalid generic task config: %s",
+			err.Error(),
+		)
+	}
+
+	// Persist the task.
+	taskID := model.NewTaskID()
+	jobID := model.NewJobID()
+	startTime := time.Now()
+	err = db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		if err := db.AddJobTx(ctx, tx, &model.Job{
+			JobID:   jobID,
+			JobType: model.JobTypeGeneric,
+			OwnerID: &genericTaskSpec.Base.Owner.ID,
+		}); err != nil {
+			return fmt.Errorf("persisting job %v: %w", taskID, err)
+		}
+
+		if err := db.AddTaskTx(ctx, tx, &model.Task{
+			TaskID:     taskID,
+			TaskType:   model.TaskTypeGeneric,
+			StartTime:  startTime,
+			JobID:      &jobID,
+			LogVersion: model.CurrentTaskLogVersion,
+		}); err != nil {
+			return fmt.Errorf("persisting task %v: %w", taskID, err)
+		}
+
+		// TODO persist config elemnts
+
+		if _, err := tx.NewInsert().Model(&model.TaskContextDirectory{
+			TaskID:           taskID,
+			ContextDirectory: contextDirectoryBytes,
+		}).Exec(ctx); err != nil {
+			return fmt.Errorf("persisting context directory files: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("persisting task information: %w", err)
+	}
+
+	logCtx := logger.Context{
+		"job-id":    jobID,
+		"task-id":   taskID,
+		"task-type": model.TaskTypeGeneric,
+	}
+	priorityChange := func(priority int) error {
+		return nil
+	}
+	if err = tasklist.GroupPriorityChangeRegistry.Add(jobID, priorityChange); err != nil {
+		return nil, err
+	}
+
+	err = task.DefaultService.StartAllocation(logCtx, sproto.AllocateRequest{
+		AllocationID:      model.AllocationID(fmt.Sprintf("%s.%d", taskID, 1)),
+		TaskID:            taskID,
+		JobID:             jobID,
+		JobSubmissionTime: startTime,
+		IsUserVisible:     true,
+		Name:              fmt.Sprintf("Generic Task %s", taskID),
+
+		SlotsNeeded:  genericTaskSpec.GenericTaskConfig.Resources.SlotsPerTask(),
+		ResourcePool: genericTaskSpec.GenericTaskConfig.Resources.ResourcePool(),
+		FittingRequirements: sproto.FittingRequirements{
+			SingleAgent: genericTaskSpec.GenericTaskConfig.Resources.IsSingleNode(),
+		},
+
+		Restore: false,
+	}, a.m.db, a.m.rm, genericTaskSpec, onAllocationExit)
+	if err != nil {
+		return nil, err
+	}
+
+	jobservice.DefaultService.RegisterJob(jobID, genericTaskSpec)
+
+	return &apiv1.CreateGenericTaskResponse{
+		TaskId:   string(taskID),
+		Warnings: pkgCommand.LaunchWarningToProto(warnings),
+	}, nil
+}
+
+func onAllocationExit(ae *task.AllocationExited) {}
```

Comment:
I am not super familiar with this new `onAllocationExit` callback, but it seems like it's not empty for all the other task types. I suspect this one may need to do something as well?

Reply:
probably stuff like add an end time for the task or rm the task context dir row.

Reply:
probably want to keep task context dir row for the webui to look at like we do for experiments or allow downloading it in the future?

Reply:
Tasks should already handle setting their own exit time or did we change that?

Reply:
Function added on line 227 above